File: src/main/scala/scorex/account/PrivateKeyAccount.java
package scorex.account;

import scorex.crypto.Crypto;
import utils.Pair;

public class PrivateKeyAccount extends PublicKeyAccount {

	private byte[] seed;
	private Pair<byte[], byte[]> keyPair;
	
	public PrivateKeyAccount(byte[] seed)
	{
		this.seed = seed;
		scala.Tuple2 pair = Crypto.createKeyPair(seed);
		this.keyPair = new Pair(pair._1(), pair._2());
		this.publicKey = keyPair.getB();
		this.address = Crypto.getAddress(this.publicKey);
	}
	
	public byte[] getSeed()
	{
		return this.seed;
	}
	
	public byte[] getPrivateKey() 
	{
		return this.keyPair.getA();
	}
	
	public Pair<byte[], byte[]> getKeyPair()
	{
		return this.keyPair;
	}
	
}

File: src/main/scala/scorex/account/PrivateKeyAccount.scala
package scorex.account

import scorex.crypto.EllipticCurveImpl
import scorex.crypto.encode.Base58
import scorex.transaction.ValidationError.GenericError

import scala.util.{Failure, Success}

sealed trait PrivateKeyAccount extends PublicKeyAccount {
  def seed: Array[Byte]

  def privateKey: Array[Byte]
}

object PrivateKeyAccount {

  private case class PrivateKeyAccountImpl(seed: Array[Byte], privateKey: Array[Byte], publicKey: Array[Byte]) extends PrivateKeyAccount

  def apply(seed: Array[Byte]): PrivateKeyAccount = {
    val pair = EllipticCurveImpl.createKeyPair(seed)
    PrivateKeyAccountImpl(seed, pair._1, pair._2)
  }

  def fromSeed(s: String): Either[GenericError, PrivateKeyAccount] = Base58.decode(s) match {
    case Success(x) => Right(PrivateKeyAccount(x))
    case Failure(e) => Left(GenericError(s"Unable to get a private key from the seed '$s': ${e.getMessage}"))
  }

}
